Extract the EMR polling code to a function.

robcowie opened the pull request:
Initial implementation.

robcowie committed:
Extract the EMR polling code to a function.

Initial implementation.

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+
+States = namedtuple('States', ['running', 'success', 'failure'])
+
+CLUSTER_STATES = States(
```
you are not currently using the CLUSTER_STATES, do we need it?

robcowie replied:
Put it there because I'd already looked them up. Happy to remove it though.

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+import boto.emr.connection
+
+
+States = namedtuple('States', ['running', 'success', 'failure'])
```
interesting use for namedtuples, the only thing to be picky is that in general to check if something is contained somewhere else sets are better than lists, and you could also write it with something like

``` python
CLUSTER_STATES = {
    'running': {'STARTING', 'BOOTSTRAPPING'},
    'success': {'TERMINATED'}
}
```

but the namedtuple is quite neat actually

robcowie replied:
Agreed.

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+    pass
+
+
+def on_cluster_status_change(conn, logger, status, prev_state):
```
this function does not use the conn

robcowie replied:
I figured I'd keep the arg signature of the callbacks identical. I haven't thought that idea through so happy to modify.

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+    logger.debug(msg)
+
+
+def on_cluster_success_status(conn, logger, status):
```
same as before

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+                 status.state, status.laststatechangereason)
+
+
+def on_cluster_failure_status(conn, logger, status):
```
and here

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+    raise EMRFailure(msg)
+
+
+def poll_for_status(conn, logger, jobflow_id,
```
more as a general comment, it's true that the process has nothing else to do at this point, but in theory this should run in another process/thread not the main one right?

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+        if status.state != prev_state:
+            on_status_change(conn, logger, status, prev_state)
+            prev_state = status.state
+
```
this is a finite state machine really, so apart from the actual EMR specifics (which can be mocked) is quite easy to test.
Would you be able to add a couple of tests for this?

AndreaCrotti commented on aws_utils/emr/polling.py:
```diff
@@ -0,0 +1,71 @@
+    """Handle EMR cluster failure status."""
+    msg = 'Jobflow failure: {} {}'.format(
+        status.state, status.laststatechangereason)
+    logger.debug(msg)
```
you might also do logger.exception here which would also include the whole stacktrace in the log.

The raise below might do the same maybe but not sure it would be the same stacktrace

Commit:
Implement Cluster Status Polling - EMR

This commit proposes a solution to wait until a cluter finishes all
its steps, fails or takes longer than expected to finish.

neebone commented on aws_utils/emr/polling_v2.py:
```diff
@@ -0,0 +1,107 @@
+
+def on_cluster_success_status(logger, status, reason):
+    """Handle EMR cluster success status."""
+    logger.debug('Cluster success: %s %s', status, reason)
```
I like to wrap these in `logger.isEnabledFor(logger.DEBUG)` pattern to avoid the text parsing, but it's not a big deal here.

eberbis replied:
I see what you mean... I wonder if this is like that and the `logger` is set to `INFO` only, then they would not get any notification at all about the status of the cluster. Perhaps it's a better option to move this up to `INFO` level, @neebone?

neebone replied:
Makes sense. Because it's cluster run health I think it deserves an INFO level logger. Agree.

alistair-broomhead replied:
Couldn't help wading in seeing as I'm getting notified, with `logger.debug('Cluster success: %s %s', status, reason)` then the string will not be formatted unless the log-level is DEBUG or below, this is the point of this form of logger call over `logger.debug('Cluster success: {} {}'.format(status, reason))`

Commit:
Set proper logger levels

Commit:
Rename EMR status checker file
